File: _Demo_and_Tests/Objective-C/AlgorithmsTests/InsertionSortTests.m
//
//  InsertionSortTests.m
//  Algorithms
//

#import <XCTest/XCTest.h>
#import "InsertionSort.h"
#import "TestsHelper.h"

@interface InsertionSortTests : XCTestCase

@end

@implementation InsertionSortTests

- (void)setUp {
    [super setUp];

    [TestsHelper reset];
}

- (void)tearDown {
    [super tearDown];
}

#pragma mark -

- (void)testAscOdd {
    [InsertionSort sort:TestsHelper.array order:InsertionSortOrderAsc];

    XCTAssertEqualObjects(TestsHelper.array, TestsHelper.sortedArrayAsc);
}

- (void)testAscEven {
    [TestsHelper.array addObject:@100];
    [TestsHelper.sortedArrayAsc addObject:@100];

    [InsertionSort sort:TestsHelper.array order:InsertionSortOrderAsc];

    XCTAssertEqualObjects(TestsHelper.array, TestsHelper.sortedArrayAsc);
}

- (void)testDescOdd {
    [InsertionSort sort:TestsHelper.array order:InsertionSortOrderDesc];

    XCTAssertEqualObjects(TestsHelper.array, TestsHelper.sortedArrayDesc);
}

- (void)testDescEven {
    [TestsHelper.array addObject:@100];
    [TestsHelper.sortedArrayDesc insertObject:@100 atIndex:0];

    [InsertionSort sort:TestsHelper.array order:InsertionSortOrderDesc];

    XCTAssertEqualObjects(TestsHelper.array, TestsHelper.sortedArrayDesc);
}

- (void)testNilArray {
    NSMutableArray *resultArray = nil;

    [InsertionSort sort:resultArray order:InsertionSortOrderAsc];

    XCTAssert(resultArray == nil);
}

- (void)testEmptyArray {
    NSMutableArray *resultArray = [NSMutableArray new];

    [InsertionSort sort:resultArray order:InsertionSortOrderAsc];

    XCTAssertEqualObjects(resultArray, @[]);
}

- (void)testArrayWithOneNumber {
    NSMutableArray *resultArray = [NSMutableArray arrayWithObject:@0];

    [InsertionSort sort:resultArray order:InsertionSortOrderAsc];

    XCTAssertEqualObjects(resultArray, @[@0]);
}

@end

File: _Demo_and_Tests/Swift/AlgorithmsTests/InsertionSortTests.swift
//
//  InsertionSortTests.swift
//  Algorithms
//

import XCTest

class InsertionSortTests: XCTestCase {

    override func setUp() {
        super.setUp()

        TestsHelper.reset()
    }

    override func tearDown() {
        super.tearDown()
    }

    // MARK: -

    func testAscOdd() {
        let resultArray = InsertionSort.sort(TestsHelper.array, order: .ASC)

        XCTAssert(resultArray ?= TestsHelper.sortedArrayAsc)
    }

    func testAscEven() {
        TestsHelper.array.append(100)
        TestsHelper.sortedArrayAsc.append(100)

        let resultArray = InsertionSort.sort(TestsHelper.array, order: .ASC)

        XCTAssert(resultArray ?= TestsHelper.sortedArrayAsc)
    }

    func testDescOdd() {
        let resultArray = InsertionSort.sort(TestsHelper.array, order: .DESC)

        XCTAssert(resultArray ?= TestsHelper.sortedArrayDesc)
    }

    func testDescEven() {
        TestsHelper.array.append(100)
        TestsHelper.sortedArrayDesc.insert(100, at: 0)

        let resultArray = InsertionSort.sort(TestsHelper.array, order: .DESC)

        XCTAssert(resultArray ?= TestsHelper.sortedArrayDesc)
    }

    func testNilArray() {
        let nilArray: [Int]? = nil
        let resultArray = InsertionSort.sort(nilArray, order: .ASC)

        XCTAssert(resultArray == nil);
    }

    func testEmptyArray() {
        let resultArray = InsertionSort.sort([] as [Int], order: .ASC)

        XCTAssert(resultArray ?= []);
    }

    func testArrayWithOneNumber() {
        let resultArray = InsertionSort.sort([0], order: .ASC)
        
        XCTAssert(resultArray ?= [0]);
    }

}
